### DISCUSSION TITLE
v2021.12.5-beta1 - Ke compensation

### DISCUSSION THREAD

**Discussion by bigbread59** [ScratMan/HASmartThermostat#50] [General]
I would like to give a small contribution in the "Ke compensation" introduced in v2021.12.5-beta1.

I found an inconsistency between the one written in
"Beta version What's new"

It is written:

**_An external component E is added to the PID output:
...
E = Ke * (target_temp - outdoor_temp)
This compensation will act like the integral of the PID_**

It seems to me that the contribution of "E" is similar to the proportional action of the contribution "P" rather than the integral "I".

Unfortunately the effect of the temperature outside the building does not propagate instantaneously inside the building but with a delay due to the thermal resistance of the walls and the thermal capacity of the internal environment (such as an RC electrical circuit).

Therefore I would like the opinion of expert peaple and open the discussion whether it is better to use:

**E = Ke * (target_temp - outdoor_temp)**
or:
**E = last_E + (Ke * (target_temp - outdoor_temp) * dt)**
similar to integral action which will be delayed as needed.

**The problem in this case is that the "E" term will grow indefinitely until the outside temperature is lower than the inside one, so a way should be found to avoid this.**

**Comment by ScratMan** [comment]
Hello, when said it works as integral, it means by adding a permanent offset to the PID output.
The difference with integral is that I is automatically tuned by the system, so the offset will depend on the current temperature ; while E does not depend on the status of the regulation: colder outside = faster losses = need more power.

The E part is directly inspired from boilers linear regulation of water output temperature based on outside temperature.

**Reply by bigbread59** [reply]
@ScratMan 
Hi,
Thanks for the clarification!
Now your implementation becomes clear and I think it can be very useful.

I'll try it.

**Reply by ScratMan** [reply]
To test it, you can look at your integral part once settled in the PID, for example mine was moving between 20 and 30 %, and then look at the `target_temp - outdoor_temp` difference (for me today it's around 22°C - 3.5°C = 18.5°C), and then compute Ke so that the integral can flow between 5 to 15%, that means the 18.5°C difference should lead to `E = 15%` on my example, it gives a `Ke = 15 / 18.5 = 0.81`

**Reply by bigbread59** [reply]
@ScratMan 
Thank you,
I just restarted the PID clearing the integrating part of the PID.
I'm waiting for PID stabilization, after that i'll do the Ke compensation.

**Comment by ScratMan** [comment]
Working fine in my bathroom...
![Screenshot_2021-12-16-17-03-53-52_c3a231c25ed346e59462e84656a70e50](https://user-images.githubusercontent.com/971926/146406535-60784194-8b25-47ef-9e5e-ebe705618630.jpg)

![Screenshot_2021-12-16-17-04-24-96](https://user-images.githubusercontent.com/971926/146406639-91801872-b3b8-47a1-b815-93475e8e6a2d.jpg)

**Reply by bigbread59** [reply]
@ScratMan 
Very good!

What type of boiler/radiators do you have?

It seem to see very short "ON/OFF" cycles.

What are your setup for following parameters?

**pwm:
min_cycle_duration:
min_off_cycle_duration:
keep_alive:
sampling_period:**

**Reply by ScratMan** [reply]
It's an electric 750W towel dryer, in a 8m² bathroom, of a 2017 house.

The configuration:
```
- platform: smart_thermostat
  name: Salle de bain
  unique_id: salle_de_bain
  heater: switch.seche_serviette
  target_sensor: sensor.popp_mold_salle_de_bain_temperature
  outdoor_sensor: sensor.fgbs222_smart_module_1_garage_temperature_2
  min_temp: 7
  max_temp: 30
  away_temp: 14
  eco_temp: 19
  comfort_temp: 22
  boost_temp: 24
  keep_alive: 00:01:00
  min_cycle_duration: 00:01:00
  precision: 0.1
  pwm : 900


```

**Reply by bigbread59** [reply]
@ScratMan 

I guess for your electrical towel dryer you are using a SSR relay

**Reply by ScratMan** [reply]
No, it's driven using  Qubino flush pilot wire module. My Home Assistant switches the module between freeze safe and comfort, and the onboard thermostat of the dryer is set to the maximum so that it doesn't switch off by itself.

**Reply by bigbread59** [reply]
@ScratMan 
Good! I learned that there is a new product! 
_**Qubino flush pilot wire module.**_ 
I'm more familiar with WIFI and zigbee products.

**Reply by ScratMan** [reply]
Unfortunately I can't find any information in English, only in French. The module doesn't appear on the Qubino web site.

https://www.amazon.fr/ZMNHJD1-Module-pilote-radiateur-%C3%A9lectrique/dp/B01HHCCVGY/ref=asc_df_B01HHCCVGY/?tag=googshopfr-21&linkCode=df0&hvadid=106800305281&hvpos=&hvnetw=g&hvrand=4566164673372812237&hvpone=&hvptwo=&hvqmt=&hvdev=m&hvdvcmdl=&hvlocint=&hvlocphy=9055445&hvtargid=pla-227166303601&psc=1

**Reply by bigbread59** [reply]
Thank you.

I took a look here:

https://qubino.com/material/Qubino_Catalogue_V_21.pdf

They are good products, more expensive than the Sonoff, Shelly and zigbee products I know and use.

**Reply by ScratMan** [reply]
They make Zwave products, while Sonoff and Shelly uses ZigBee or WiFi. Zwave is usually more expensive, the standard is not as open as ZigBee but it's more robust and secured.

**Reply by bigbread59** [reply]
I will keep Zwave products in mind for "safety critical" functions/applications.

**Reply by ScratMan** [reply]
Yeah, ZigBee is good for "standard" home appliances, like shutters, lights, or sockets.
For openings detection, alarms or heating I prefer the flexibility, security and robustness of Zwave.

**Reply by bigbread59** [reply]
I agree

**Reply by bigbread59** [reply]
Which Z-Wave gateway do you use?
Do you have some suggestion?

**Reply by ScratMan** [reply]
I'm using an Aeotec stick on a Raspberry pi.

**Comment by nagyrobi** [comment]
Question:
The doc says _The Ke gain depends on the insulation of the building, on recent buildings with good insulation, a gain of 0.6 is recommended._ 
What's a good value to start with poor uninsulated older buildings? What are min/max values? Is it a linear value?
Like 0.01 poor insulated and 0.99 best insulated? Or reverse?

**Reply by nagyrobi** [reply]
Anybody a hint please?

**Reply by ScratMan** [reply]
Sorry for late reply, to determine the value, please check the pid_i value when settled, and use method described in this post above to compute your base Ke.
https://github.com/ScratMan/HASmartThermostat/discussions/50#discussioncomment-1822781
